src/System.Linq/src/System/Linq/Partition.cs

```diff
@@ -389,5 +389,245 @@ public int GetCount(bool onlyIfCheap)
                 return Count;
             }
         }
+
+        private sealed class EnumerablePartition<TSource> : Iterator<TSource>, IPartition<TSource>
+        {
+            private readonly IEnumerable<TSource> _source;
+            private readonly int _minIndexInclusive;
+            private readonly int _maxIndexInclusive;
+            private IEnumerator<TSource> _enumerator;
+
+            internal EnumerablePartition(IEnumerable<TSource> source, int minIndexInclusive, int maxIndexInclusive)
+            {
+                Debug.Assert(source != null);
+                Debug.Assert(!(source is IList<TSource>), $"The caller needs to check for {nameof(IList<TSource>)}.");
+                Debug.Assert(minIndexInclusive >= 0);
+                Debug.Assert(maxIndexInclusive >= 0);
+                Debug.Assert(minIndexInclusive <= maxIndexInclusive);
+
+                _source = source;
+                _minIndexInclusive = minIndexInclusive;
+                _maxIndexInclusive = maxIndexInclusive;
+            }
+
+            public override Iterator<TSource> Clone()
+            {
+                return new EnumerablePartition<TSource>(_source, _minIndexInclusive, _maxIndexInclusive);
+            }
+
+            public int GetCount(bool onlyIfCheap)
+            {
+                if (onlyIfCheap)
+                {
+                    return -1;
+                }
+
+                int count = _source.Count();
+                if (count <= _minIndexInclusive)
+                {
+                    return 0;
+                }
+
+                return Math.Min(count - 1, _maxIndexInclusive) - _minIndexInclusive + 1;
+            }
+
+            public override bool MoveNext()
+            {
+                // Need to handle -1 and 0 explicitly since we have a default: case.
+                if (_state <= 0)
+                {
+                    Dispose();
+                    return false;
+                }
+
+                switch (_state)
+                {
+                    case 1:
+                        _enumerator = _source.GetEnumerator();
+                        _state = 2;
+                        goto case 2;
+                    case 2:
+                        if (!SkipBeforeFirst(_enumerator))
+                        {
+                            // Reached the end before we finished skipping.
+                            break;
+                        }
+
+                        _state = 3;
+                        goto default;
+                    default:
+                        int taken = _state - 3;
+                        if (taken <= _maxIndexInclusive - _minIndexInclusive && _enumerator.MoveNext())
+                        {
+                            _state++;
```

Comment:
Doesn't using _state like this cause a problem for really long enumerables with > int.MaxValue - 3 elements?